src/bitcoin-cli.cpp

```diff
@@ -291,6 +293,184 @@ class GetinfoRequestHandler: public BaseRequestHandler
     }
 };
 
+/** Process netinfo requests */
+class NetinfoRequestHandler : public BaseRequestHandler
+{
+private:
+    bool IsAddrIPv6(const std::string& addr) const
+    {
+        return !addr.empty() && addr.front() == '[';
+    }
+    bool IsInboundOnion(int mapped_as, const std::string& addr, const std::string& addr_local) const
+    {
+        return mapped_as == 0 && addr.find("127.0.0.1") == 0 && addr_local.find(".onion") != std::string::npos;
+    }
+    bool IsOutboundOnion(const std::string& addr) const
+    {
+        return addr.find(".onion") != std::string::npos;
```

Comment:
I haven't seen this to be an issue yet. Happy to look at a proposal.

Reply:
I see this for an IPv4 connection to www.onionfoo.com:8345, mistakenly labelled as "conn=onion" instead of "conn=ipv4".

```
<-> relay  conn minping   ping lastsend lastrecv uptime id version               address
out  full onion       4      4        0        0      0  0 70015/Satoshi:0.20.0/ www.onionfoo.com:8345
```

```suggestion
        const char* suffix = ".onion";
        const size_t suffix_len = 6;
        return addr.length() > suffix_len &&
               addr.compare(addr.length() - suffix_len, std::string::npos, suffix) == 0;
```

Reply:
Thanks, will test; can pull in a commit from you in the follow-up if this is merged. If you see a way to distinguish between inbound ipv4 and ipv46 conns, that would be great too.

Reply:
In more than a month of running this and also the original python script, I still have not seen a case like this and have not found the node in question on the network. Seems like a pretty pathological edge case but happy to look at tightening the detection in the follow-up.

Reply:
Simpler yet, maybe just begin the outbound onion check with `mapped_as == 0` like for the inbound onion check. Still haven't seen an actual issue though.

Reply:
I have still not yet encountered the described edge case above, but it should be noted that replacing `find` with "string ends with" causes many outbound onions to not be detected when they end with ".onion:<port>", e.g. ".onion:8333".

Reply:
I have still not yet encountered the described edge case above, but it should be noted that the proposal to replace `find` with "string ends with" causes many outbound onions to not be detected when they end with ".onion:<port>", e.g. ".onion:8333".

Reply:
I may try something like `ends_with ".onion" || finds ".onion:" x chars from the end`

Reply:
I may try something like `ends_with ".onion" || finds ".onion:" x chars from the end` in my dev branch, but it's an edge case that I haven't seen yet and doesn't seem necessary to do it here.

Reply:
This seems to cover everything:
```cpp
bool IsOutboundOnion(const std::string& addr, int mapped_as) const
{
    const size_t onion_len{ONION.length()};
    const size_t pos{addr.rfind(ONION)};
    return mapped_as == 0 && (pos == addr.size() - onion_len || pos == addr.find_last_of(":") - onion_len);
}
```